internal/provider_framework/attribute_validators.go

```diff
@@ -0,0 +1,219 @@
+	tflog.Debug(ctx, "Validating attribute is a URL with acceptable scheme", map[string]interface{}{
+		"attribute":         attrPathToString(req.AttributePath),
+		"acceptableSchemes": strings.Join(av.acceptableSchemes, ","),
+	})
+
+	var v types.String
+	diags := tfsdk.ValueAs(ctx, req.AttributeConfig, &v)
+	if diags.HasError() {
+		res.Diagnostics.Append(diags...)
+		return
+	}
+
+	if v.Null || v.Unknown {
+		return
+	}
+
+	u, err := url.Parse(v.Value)
+	if err != nil {
+		res.Diagnostics.AddAttributeError(
+			req.AttributePath,
+			"Invalid URL",
+			fmt.Sprintf("Parsing URL %v failed: %+v", v, err),
+		)
+		return
+	}
+
+	if u.Host == "" {
+		res.Diagnostics.AddAttributeError(
+			req.AttributePath,
+			"Invalid URL",
+			fmt.Sprintf("URL %v contains no host", v),
+		)
+		return
+	}
+
+	for _, s := range av.acceptableSchemes {
+		if u.Scheme == s {
+			return
+		}
+	}
+
+	res.Diagnostics.AddAttributeError(
+		req.AttributePath,
+		"Invalid URL",
+		fmt.Sprintf("URL scheme expected to be one of %q, but is: %v", av.acceptableSchemes, u.Scheme),
+	)
+	return
+}
+
+// requiredWithAttributeValidator checks that a set of *tftypes.AttributePath,
+// including the attribute it's applied to, are set simultaneously.
+// This implements the validation logic declaratively within the tfsdk.Schema.
+//
+// The provided tftypes.AttributePath must be "absolute",
+// and starting with top level attribute names.
+type requiredWithAttributeValidator struct {
+	requiredAttributes []*tftypes.AttributePath
+}
+
+// requiredWith is a helper to instantiate requiredWithAttributeValidator.
+func requiredWith(attributePaths ...*tftypes.AttributePath) tfsdk.AttributeValidator {
+	return &requiredWithAttributeValidator{attributePaths}
+}
+
+var _ tfsdk.AttributeValidator = (*requiredWithAttributeValidator)(nil)
+
+func (av requiredWithAttributeValidator) Description(ctx context.Context) string {
+	return av.MarkdownDescription(ctx)
+}
+
+func (av requiredWithAttributeValidator) MarkdownDescription(ctx context.Context) string {
+	return fmt.Sprintf("Ensure that if an attribute is set, also these are set: %q", av.requiredAttributes)
+}
+
+func (av requiredWithAttributeValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, res *tfsdk.ValidateAttributeResponse) {
+	tflog.Debug(ctx, "Validating attribute is set together with other required attributes", map[string]interface{}{
+		"attribute":          attrPathToString(req.AttributePath),
+		"requiredAttributes": av.requiredAttributes,
+	})
+
+	var v attr.Value
+	diags := tfsdk.ValueAs(ctx, req.AttributeConfig, &v)
+	if diags.HasError() {
+		res.Diagnostics.Append(diags...)
+		return
+	}
+
+	for _, path := range av.requiredAttributes {
+		var o attr.Value
+		req.Config.GetAttribute(ctx, path, &o)
+
+		isVNull, diags := isAttrValueNull(ctx, req.AttributePath, v)
```

Comment:
Looks like the determination of whether `v` is null, is performed on each iteration. Believe this could be moved above the `for` loop as the return value will be stable.